oasislmf/cli/model.py

```diff
@@ -280,6 +281,10 @@ def action(self, args):
             inputs.get('complex_lookup_config_file_path', required=False, is_path=True),
             'Complex lookup config JSON file path', preexists=False
         )
+        user_data_dir = as_path(
+            inputs.get('user_data_path', required=False, is_path=True),
+            'Directory containing additional user-supplied model data files', preexists=False
+        )
```

Comment:
Any idea why the duplication between the CLI argument parser and here? The help string and `required` appear in both places.

Reply:
Not particularly. Another one to keep in mind is that there's a slightly confusing swap from`names-like-this` and `names_like_this` - even though the CLI defines arguments as the former, the `inputs` contains them as the latter.

Reply:
What is the duplication here? The `inputs` object (created from the `cli.base.InputValues` class) is a more user-friendly object wrapper around the arguments `args` object parsed by the command parser (when `action` is called), and `as_path` is a "path cleaner" that is applied to the path-like arguments in `inputs`. There is some overlap between some of the optional arguments used by the parser, the `inputs` object and the `as_path` cleaner, because this is required in the way that `InputValues` and the path cleaner are meant to work.

Reply:
Also, with respect to the format of the argument option names for command arguments, as you may know, from using various command line tools, e.g. `curl`, the convention is hyphen-separated, e.g.

    $ curl --help
    Usage: curl [options...] <url>
    Options: (H) means HTTP/HTTPS only, (F) means FTP only
         --anyauth       Pick "any" authentication method (H)
         -a, --append        Append to target file when uploading (F/SFTP)
         ...

but the variables which are used internally by the command line tool to refer to those option names don't need to be hyphen separated. There is no rule saying that they have to use the same hyphen-separated convention, which is probably better for readability. Internally, the variable name syntax for CLI argument option names in the MDK follows Python convention, which is underscoring. Argparse automatically converts hyphen-separated option names to values where the hyphen is replaced with an underscore, e.g. `keys-file-path` -> `keys_file_path`, etc.

Reply:
I believe Carl was referring to the fact that the help string, whether an option is required, etc., are defined both in `add_args` and `action`. Am I right to think that the ones in `action` are used in error messages (e.g. "XXX was indicated as pre-existing...")?

Interesting point about argparse doing the conversion - I didn't realise that was actually its default behaviour.

Reply:
The help text in `add_args` is used when you run `--help` on the command. In `action` we keep it because it may be useful when raising an exception for example.

Reply:
In Python you can't have hyphenated variable names anyway, and so if you the convention of hyphenated command line option names then you must convert those names to a format that can be processed by Python, e.g. underscoring.

Reply:
Oh, yes, that's an excellent point - that does make it obvious that argparse has to do a conversion.

Reply:
I'm just pointing out that when you add an argument you have to do it in at least two places and that could be considered a code smell. If there is no better alternative to the way it's currently done, that's fine.

Reply:
There are probably better alternatives, but for the moment we will keep it as it is. Help text doesn't figure at all in the logic of processing the commands, so the duplication isn't going to break anything, therefore it can't be considered code smell.